lib/c-api/src/wasm_c_api/wasi/mod.rs

```diff
@@ -344,7 +345,98 @@ pub unsafe extern "C" fn wasi_get_wasi_version(module: &wasm_module_t) -> wasi_v
         .unwrap_or(wasi_version_t::INVALID_VERSION)
 }
 
-/// Takes ownership of `wasi_env_t`.
+/// Non-standard type wrapping `wasm_extern_t` with the addition of
+/// two `wasm_name_t` respectively for the module name and the name of
+/// the extern (very likely to be an import). This non-standard type
+/// is used by the non-standard `wasi_get_unordered_imports` function.
+#[allow(non_camel_case_types)]
+#[derive(Clone)]
+pub struct wasm_named_extern_t {
+    module: Box<wasm_name_t>,
+    name: Box<wasm_name_t>,
+    r#extern: Box<wasm_extern_t>,
+}
+
+wasm_declare_boxed_vec!(named_extern);
+
+/// Non-standard function to get the module name of a
+/// `wasm_named_extern_t`.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_module(named_extern: &wasm_named_extern_t) -> &wasm_name_t {
+    named_extern.module.as_ref()
+}
+
+/// Non-standard function to get the name of a `wasm_named_extern_t`.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_name(named_extern: &wasm_named_extern_t) -> &wasm_name_t {
+    named_extern.name.as_ref()
+}
+
+/// Non-standard function to get the wrapped extern of a
+/// `wasm_named_extern_t`.
+#[no_mangle]
+pub extern "C" fn wasm_named_extern_unwrap(named_extern: &wasm_named_extern_t) -> &wasm_extern_t {
+    named_extern.r#extern.as_ref()
+}
+
+/// Non-standard function to get the imports needed for the WASI
+/// implementation with no particular order. Each import has its
+/// associated module name and name, so that it can be re-order later
+/// based on the `wasm_module_t` requirements.
+///
+/// This function takes ownership of `wasm_env_t`.
+#[no_mangle]
+pub unsafe extern "C" fn wasi_get_unordered_imports(
+    store: &wasm_store_t,
+    module: &wasm_module_t,
+    wasi_env: &wasi_env_t,
+    imports: &mut wasm_named_extern_vec_t,
+) -> bool {
+    wasi_get_unordered_imports_inner(store, module, wasi_env, imports).is_some()
+}
+
+unsafe fn wasi_get_unordered_imports_inner(
+    store: &wasm_store_t,
+    module: &wasm_module_t,
+    wasi_env: &wasi_env_t,
+    imports: &mut wasm_named_extern_vec_t,
+) -> Option<()> {
+    let store = &store.inner;
+
+    let version = c_try!(
+        get_wasi_version(&module.inner, false).ok_or_else(|| CApiError {
+            msg: "could not detect a WASI version on the given module".to_string(),
+        })
+    );
+
+    let import_object = generate_import_object_from_env(store, wasi_env.inner.clone(), version);
+
+    *imports = import_object
+        .into_iter()
+        .map(|((module, name), export)| {
+            let module = Box::new(module.into());
+            let name = Box::new(name.into());
+            let extern_inner = Extern::from_vm_export(store, export);
+
+            Box::new(wasm_named_extern_t {
+                module,
+                name,
+                r#extern: Box::new(wasm_extern_t {
+                    instance: None,
+                    inner: extern_inner,
+                }),
```

Comment:
TODO: review `instance: None`, aren't `wasm_extern_t`s callable? is `instance: None` a problem for data being passed to the user?

Reply:
We don't have an access to `Instance` here since we read the externs from `Module`.

`wasm_extern_t` is defined as:

```rust
pub struct wasm_extern_t {
    // this is how we ensure the instance stays alive
    pub(crate) instance: Option<Arc<Instance>>,
    pub(crate) inner: Extern,
}
```

We don't need to ensure that `Instance` stays alive, it doesn't exist at this step :-).